transactions/read-only/parallel.md

```diff
@@ -0,0 +1,215 @@
+| get                         |  none  | yes                    |                        |
+
+### Net Messages
+Net messages can be classified into sync and non-sync:
+- Non-sync messages do not modify states.
+- Sync messages are signed_block, packed_transaction. They may modify states.
+
+| Message                     | Global data modified            | threads involved            | Safe to read-only trx?|
+|-----------------------------|---------------------------------|-----------------------------|-----------------------| 
+| handshake                   | none                            | net, main (handshake check) | yes        |
+| go_away                     | none                            | net              | yes                   |
+| time                        | none                            | net              | yes                   |
+| notice                      | none                            | net              | yes                   |
+| request                     | none                            | net, main (controller::fetch_block_by_id()) | yes   |
+| sync_request                | none                            | net, main (controller::fetch_block_by_number()) | yes |
+| packed_transaction          | chainbase, forkdb, producer, net, controller | net, producer, main | no    |
+| signed_block                | chainbase, forkdb, producer, net, controller | net, producer, main | no    |
+
+### SHiP
+SHiP receives blockchain state data, saves it to files, and sends data to nodes who request it.
+
+| Request               |  Data modified                  | Safe to read-only trx? |
+|-----------------------|---------------------------------|------------------------|
+| get_status            |  internal                       | yes                    |
+| get_blocks            |  internal                       | yes                    |
+| get_blocks_ack        |  internal                       | yes                    |
+
+## Design Decisions
+
+### Window Toggling
+To support toggling between `read` and `write` windows, configurable options `read-window-time`, `write-window-time`, and `read-only-max-queued-time` are provided. 
+- From `write` to `read`: When a new read-only transaction is queued, if the time the earliest transaction has been queued exceeds `read-only-max-queued-time`; or at the end of the `write` window and read-only transaction queue not empty.
+- From `read` to `write`: when read-only transaction queue is emptied by the read-only threads; or at the end of `read` window.
+
+In the following state diagram, `longest_queued_time = now - the time when the oldest trx was queued`, `write_window_deadline = time when write window starts + write-window-time`, and `read_window_deadline = time when read window starts + read-window-time`
+
+```mermaid
+flowchart TD
+    A(((write window))) -->|push a new trx| B[longest_queued_time < read-only-max-queued-time?]
```

Comment:
Should we check `longest_queued_time < read-only-max-queued-time` only when a new read-only transaction is queued? This means that if we don't receive read only transaction for a while, we may not process them for a long time.

Reply:
We have `write-window-time` to guard that. We would switch to `read` window when `write-window-time` has passed.

Reply:
Malse sense, thanks!